### ISSUE TITLE
Horovod trained models are not identical across workers

### ISSUE THREAD

**Issue by sudheerExperiments** [horovod/horovod#854]
Hello @alsrgv,

I am training a CNN model using Adam optimizer on a Multi GPU Multi node cluster with 2 compute nodes, each node has 4 GPUs, so in total I have 8 workers. 

I need a clarification about something. From what I have read Horovod works in this way. Correct me if I am wrong. 1) Each process loads different chunk of data from the dataset for training. 2) At the end of each epoch, Horovod aggregates local gradients from workers, averages the gradients and broadcasts the new gradients to all other workers, so in the end, all workers will have identical copy of gradients. 3) The workers then use the updated copy to train for n no. of epochs.

Similar explanation is provided by @tgaddair in the below issue.

https://github.com/horovod/horovod/issues/820#issuecomment-462101048

Since all workers have identical copy of weights, we just have to save the weights of rank 0 and use it for inference. When I used the weights from rank 0, I noticed something odd, the prediction performance is very bad. So I saved weights from all models to see what's happening, surprisingly, the models saved on different workers gave me different accuracy, the model from worker 0 performance is bad compared to models from other workers?

Is this expected behavior or am I doing something wrong? Can you please explain me why its happening please? 

P.S. I am using Horovod 16.0, Keras 2.2.4 on Tensorflow gpu 1.12.0. I am loading different chunks of data as follows.

```python
# Load data
# Data is per-processed into chunks, so that every process loads different copy of data.
filename = '/home/cluster/Documents/' + 'data_chunk_' + str(hvd.rank() + 1) + '.npy'
data = np.load(filename)
filename_1 = '/home/cluster/Documents/' + 'labels_chunk_' + str(hvd.rank() + 1) + '.npy'
labels = np.load(filename_1)
```
Then I followed all the steps described in 'keras_mnist_advanced.py' example.

My dataset has 2150 instances divided into 8 chunks. I used batch size of 32, so in total each worker gets data in 3 mini-batches while training. I called the fit function as follows.

```python
# I used BroadcastGlobalVariablesCallback, MetricAverageCallback, and LearningRateWarmupCallback in callbacks
model.fit(data, labels, verbose=1, epochs=500, batch_size=32, callbacks=callbacks)
``` 

Thanks and Regards,
Sudheer.

**Comment by sudheerExperiments** [comment]
> Hello @alsrgv,
> 
> I am training a CNN model using Adam optimizer on a Multi GPU Multi node cluster with 2 compute nodes, each node has 4 GPUs, so in total I have 8 workers.
> 
> I need clarification with this. I think this is how Horovod works. Correct me if I am wrong. 1) Each process loads different chunk of data from the dataset for training. 2) At the end of each epoch, Horovod calculates local gradients for each worker, averages the gradients and broadcasts the new gradients to all other workers, so at the end of this process, all workers have identical copy of weights. 3) All the workers then use the updated copy to train for n no. of epochs.
> 
> If I am right then at the end of the training, every worker will have same copy of model. However, I noticed something odd. The models that are saved across the workers are not same, they are giving me different predictions. The accuracy is worse with some copies of models than the others.
> 
> Is this expected behavior or am I doing wrong? Can you please explain why this is happening please?
> 
> I implemented my code as per 'keras_mnist_advanced.py' code in Horovod examples.
> 
> P.S. I am using Horovod 16.0, Keras 2.2.4 on Tensorflow gpu 1.12.0.
> 
> Thanks and Regards,
> Sudheer.

**Comment by tgaddair** [comment]
Hey @sudheerExperiments, similar to my suggestion in #820, could you try broadcasting model parameters periodically and see if that solves the model divergence issue and improves accuracy?  It may be that there's something else causing your issue in particular, but if it's also due to something like numerical stability, then that may solve your issue as well.

**Comment by sudheerExperiments** [comment]
@tgaddair, Could you please provide me an example on how to use broadcast? 

I searched on google but didn't find anything. 

With Regards, 
Sudheer.
